proposals/0013-expanded-comparison-sampling.md

```diff
@@ -0,0 +1,441 @@
+ whether the expanded comparison sample methods indicated here are supported
+ by the given hardware and runtime.
+
+## Testing
+
+### Correct Behavior Testing
+
+Verify that `SampleCmpGrad` and the new `CalculateLevelOfDetail*` overloads
+ generate appropriate code including the new DXIL and SPIRV intrinsics for all
+ shader stages.
+Verify that `SampleCmpBias` produces appropriate code including the new DXIL and
+ SPIRV intrinsics for pixel, compute, and node shader stages.
+Where the `DerivativesInMeshAndAmplificationShadersSupported` capability bit is
+ supported, verify that `SampleCmpBias` produces appropriate code including the
+ new DXIL and SPIRV intrinsics for mesh and amplification shader stages.
+For all of the above, verify that no validation errors are produced for the
+ generated code.
+For all of the above, wherever `SampleCmpGrad` and `SampleCmpBias` are used,
+ check that the appropriate shader flag is set.
+The above should all be doable with LIT filecheck tests.
+
+#### Diagnostics Testing
+
+Use `SampleCmpBias` in code reachable from the entry point of each unsupported
+ shader stage (vertex, geometry, tesselation stages, ray tracing stages, and
+ mesh stages if lacking hardware support for mesh derivatives)
+Using `SampleCmpBias`, `SampleCmpGrad`, and the new `CalculateLevelOfDetail*`
+ overloads in an unsupported shader model on a supported stage and verify that
+ a special error indicating the require version is produced.
+
+### Validation Testing
+
+Test that the Bias range limits tested by `SampleBiasFail` are respected by
+ `dx.op.sampleCmpBias`. (see Instr.ImmBiasForSampleB)
+Test that DXIL validation of allowed resource types must be performed for
+  `dx.op.sampleCmpBias` and `dx.op.sampleCmpGrad`:
+
+* Instr.SampleCompType - texture resources are declared with types that are
+  sampleable in the current environment.
+* Instr.SamplerModeForSampleC - sampler resource used by `SampleCmpBias` and
+  `SampleCmpGrad` are of type `SamplerComparisonState`.
+* Instr.ResourceKindForCalcLOD  - texture resources used by
+  CalculateLevelOfDetail* are among the valid texture objects.
+* Instr.ResourceKindForSample  - texture resources used by
+  `SampleCmpBias` and `SampleCmpGrad` are among the valid texture objects.
+* Instr.ResourceClassForSamplerGather - texture resources used by
+  `SampleCmpBias` and `SampleCmpGrad` must be SRVs.
+
+### Execution Testing
+
+Since there is no defined equation to determine which mip level an
+ implementation should use, the best approach is to compare the results of
+ different operations involving mip levels to ensure that they are consistent.
+
+To test `CalculateLevelOfDetail*` methods,
+use a texture with sentinel values in each mip level and ensure that the mip
+  level returned by CalculateLevelOfDetail matches the behavior of the original
+  SampleCmp operation by providing the sentinel value of the mip level
+  CalculateLevelOfDetail* made you expect and making sure that it matches.
```

Comment:
It might be just me, but I found the wording here confusing.  I feel like a simpler step-by-step breakdown might have been easier to follow.  Here's how I interpreted this (eventually):
- Fill each mip level of a texture with a single value matching the mip level, such as `0` for mip level 0, `1` for mip level 1, and so on.
- In a shader, perform a CalculateLevelOfDetail operation to retrieve the expected mip level for a sample operation being tested.
- With comparison mode set to equal on the comparison sampler, perform a SampleCmp operation, providing the expected mip level as the comparison value.
- Return the comparison result, which is expected to be 1 for success.

Reply:
I'll take another run at it with your comments

Reply:
Done